### ISSUE TITLE
Assets not loading on installation via CocoaPods

### ISSUE THREAD

**Issue by rafaelks** [helpscout/beacon-ios-sdk#91]
**Installation:** CocoaPods, version 2.0.0

**Actual outcome**

No asset is being presented on the beacon controller, and on console we have these issues:

> Could not load the "close" image referenced from a nib in the bundle with identifier "net.helpscout.beacon"
> Could not load the "back" image referenced from a nib in the bundle with identifier "net.helpscout.beacon"

**Expected outcome**

Images should load.

**Comment by ScottPetit** [comment]
Hey @rafaelks can you give me a bit more info about your project setup?

1. Which version of Cocoapods are you using? `pod --version`
2. Are you including `use_frameworks!` in your Podfile?
3. Is this for a native iOS project or for a React Native or other project?

**Comment by rafaelks** [comment]
@ScottPetit 

1. `1.10.0.beta.2`
2. Yes
3. Native iOS project (Swift only)

**Comment by ScottPetit** [comment]
@rafaelks Thanks for the info.  I'm going to take a look at what's going on with `1.10.0.beta.2` but in the meantime Beacon should work fine on the latest stable release `1.9.3` if you can rollback to that.

**Comment by rafaelks** [comment]
@ScottPetit It still doesn't work properly when using 1.9.3. Background shows correctly for docs screen, but not this one (screenshot). Also, no asset is displayed.

![Simulator Screen Shot - iPhone 8 (English) - 2020-08-25 at 12 16 55](https://user-images.githubusercontent.com/551004/91192927-e7821400-e6cc-11ea-9a0e-79eb91fc474c.png)
![Simulator Screen Shot - iPhone 8 (English) - 2020-08-25 at 12 16 52](https://user-images.githubusercontent.com/551004/91192936-e9e46e00-e6cc-11ea-9543-d03ce1e1240b.png)

**Comment by ScottPetit** [comment]
@rafaelks yea I just ran a sample app on `1.10.0.beta.2` without being able to generate that error.  Can you verify Beacon has a `Beacon.bundle` in the Pods directory (e.g.) 
<img width="285" alt="Screen Shot 2020-08-25 at 10 21 11 AM" src="https://user-images.githubusercontent.com/2100009/91193617-eba63580-e6bc-11ea-815a-03885f0381b5.png">

Also any other things about your machine that could be different, like beta macOS, beta Xcode (I verified I'm not seeing this on my machine with Xcode 12.0b5).

Also another potential fix could be deleting the Beacon cache with Cocoapods `pod cache clean Beacon`, deleting the Pods directory and reinstalling.

**Comment by rafaelks** [comment]
@ScottPetit I removed the entire Pods folder, all the cache from CocoaPods, and also cleaned the `DerivedData` -- it remains with the same problem.

I'm using Xcode is 11.6 (stable version) and macOS Catalina 10.15.6.

I can see the Beacon.bundle, yes.

<img width="399" alt="Screen Shot 2020-08-25 at 12 48 00" src="https://user-images.githubusercontent.com/551004/91197527-c8d24c00-e6d1-11ea-843a-cf42213635a4.png">

**Comment by ScottPetit** [comment]
Hey @rafaelks I'm really sorry for the back and forth but this is definitely a confusing issue.  Would you mind sharing your Podfile, either here or over [support](mailto:support@helpscout.com) for me to take a look at and see if there's something wrong with our podspec that's causing issues with your configuration?

**Comment by rafaelks** [comment]
@ScottPetit Sure! I've shared in our internal open ticket, asked the person in charge to forward to you. Thanks!

**Comment by ScottPetit** [comment]
Hey @rafaelks unfortunately I don't have a solution just yet.  A colleague and I both created a project with the pod file you provided (thanks for that) and assets worked as expected.  It seems like based on your other issue you maybe integrated Beacon through a different method which would be my current suggestion.

There's also the possibility that not using `use_frameworks!` could resolve the issue (although it could present issues with other Pods).

I'll keep on taking a look to see if I've got any other solutions, and reach out when there's a new version to see if that resolves it.

**Comment by rafaelks** [comment]
@ScottPetit Can you take a look at this response? I checked this Podspec and there is no usage of `resource_bundle` as well: https://stackoverflow.com/a/59372324/12182590.

**Comment by ScottPetit** [comment]
Hey @rafaelks, so yea our Podspec uses `resources` instead of `resource_bundles`.  The Podspec documentation isn't entirely clear but my understanding from the names was that if you provided a `.bundle` as part of your framework then you should use `resources` vs `resource_bundles` (A `.bundle` file is even used in the example for `resources`.).

That said I created an updated Podspec locally and if you're willing and have a bit of time to check it on your machine that'd be much appreciated.  I'll lay out the steps below:

1. Clone this repository (probably so it shares the same parent directory as the project in question)
2. Replace Beacon.podspec in your newly cloned directory with the contents [here](https://gist.github.com/ScottPetit/347c994e184f939e12bf7bed34772582)
3. Update the pod file in your app to point at your new local Podspec.  If they have the same parent directory then that should look something like `pod 'Beacon', :path => '../beacon-ios-sdk/Beacon.podspec'`
4. If Cocoapods says you're using version 2.0.1 of Beacon then it should be using Cocoapods `resource_bundles`
5. Report back if it fixes the asset problem.

I totally get if you don't want to go through all the hassle and really appreciate you helping so much already.

**Comment by rafaelks** [comment]
@scottyab Hey Scott, thanks for the explanation! I tried that, and it still have the same issues. 😥

Any other ideas?

**Comment by ScottPetit** [comment]
@rafaelks Thanks for giving it a try.  Umm one last attempt could be upgrading to 2.0.1, although it doesn't have any specific updates to the asset loading perhaps updating is enough to get it to work.

**Comment by rafaelks** [comment]
@ScottPetit Unfortunately that didn't work. I noticed one thing tho... on the script that CocoaPods generates to embed all frameworks, Beacon is the only one using a different path. Do you know why is that?

(this is automatically generated by CocoaPods in build phases)

<img width="1069" alt="Screen Shot 2020-08-31 at 14 38 15" src="https://user-images.githubusercontent.com/551004/91749357-bb1c3b00-eb97-11ea-9f28-7c9e17ba1c2d.png">

**Comment by ScottPetit** [comment]
@rafaelks I can't say for certain, but pretty likely is only because Beacon is distributed as a framework.  So Cocoapods is building all of your other projects source files and compiling them into a framework, whereas Beacon uses `vendored_frameworks` [here](https://github.com/helpscout/beacon-ios-sdk/blob/29f59ce34cc99f7457eab916cee292ca38b7e22e/Beacon.podspec#L20) to provide a framework that we've already built.

**Comment by rafaelks** [comment]
@ScottPetit Got it, makes sense. FWIW I tried also copying the framework into my project and using it outside CocoaPods, and I had the exact same issue. 😥

I created a whole project demo with the CocoaPods installation for me and it works just fine. Will keep investigating here and seeing if I can find the issue.

**Comment by AyushBherwani1998** [comment]
>     1. Which version of Cocoapods are you using? `pod --version`
1.9.3
>     2. Are you including `use_frameworks!` in your Podfile?
Yes
>     3. Is this for a native iOS project or for a React Native or other project?
Flutter project, using MethodChannels to open the Beacon

Hey @ScottPetit I'm facing the similar issue. The close button, back button is not getting rendered. Also, the background is transparent.

**Comment by AyushBherwani1998** [comment]
It works fine for 1.0.5 version, after 2.0.0 the assets stop loading.

**Comment by jotaemepereira** [comment]
Hey @AyushBherwani1998! Are you installing the SDK via CocoaPods, if that's the case, do you mind sharing your Podfile? In the meantime, I've created this [ticket](https://helpscout.atlassian.net/browse/IOSSDK-399) to have to track this internally. Thanks!

**Comment by AyushBherwani1998** [comment]
Hey @jotaemepereira, thank you for responding. I'm installing the SDK via CocoaPods. [Checkout the Podfile](https://gist.github.com/AyushBherwani1998/ee83f624a71892e14a5ad8a9910f9d79).

Also, I tried to integrate the SDK in a fresh Flutter project and it works perfectly fine.

**Comment by ScottPetit** [comment]
Hey @AyushBherwani1998 👋,

So full transparency here, we're not exactly sure why this is happening.  As you've seen this issue can crop up in some projects but not in others.  My suspicion is it's an issue with how Cocoapods is handling multi language (We added our first Swift code in 2.0.0) frameworks that bundle their own resources.  That said we'll have a release coming out very shortly that will change our Podspec somewhat substantially (by introducing XCFramework support) that could potentially resolve this issue for you.  

Another option would be to manually link the Beacon.framework, this process will be improved as we introduce XCFramework support, but could also help you here.  We don't have documented steps for manually integrating in a Flutter project, but possibly [this](https://flutter.dev/docs/development/ios-project-migration) could be of assistance if you decide to go that route.

**Comment by AyushBherwani1998** [comment]
No, thanks for looking into it. I'll try to manually link the Beacon.framework and see if it works out.

**Comment by ScottPetit** [comment]
Hey @rafaelks & @AyushBherwani1998,

I just wanted to let y'all know that we've released a new version (2.1.0) that updates Beacon to be distributed as an `.xcframework`.  In doing this we made some updates to our `Podspec` to accommodate this which maybe, hopefully, 🤞, could resolve this issue for y'all.  

Since we've had such a hard time reproducing this locally if y'all wouldn't mind updating to that version and reporting back if it made any difference in your setup.  Thanks.

**Comment by rafaelks** [comment]
@ScottPetit Thanks! Assets are loading normally for me now, but I found this problem:

![Simulator Screen Shot - iPhone 11 - 2020-10-30 at 08 40 40](https://user-images.githubusercontent.com/551004/97701130-b1299200-1a8b-11eb-9733-7b3eca4579ba.png)

There is no way for the user to leave this screen after sending a message... it needs to close the app. Is there a way to resolve this problem in code?

**Comment by AyushBherwani1998** [comment]
Hey @ScottPetit, after updating pod, the app crashes on opening the Beacon with this exception
```
Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '*** -[__NSPlaceholderDictionary initWithObjects:forKeys:count:]: attempt to insert nil object from objects[0]'
*** First throw call stack:
(
	0   CoreFoundation                      0x00007fff2043a126 __exceptionPreprocess + 242
	1   libobjc.A.dylib                     0x00007fff20177f78 objc_exception_throw + 48
	2   CoreFoundation                      0x00007fff204b6cdb _CFThrowFormattedException + 194
	3   CoreFoundation                      0x00007fff204c21f8 -[__NSPlaceholderDictionary initWithObjects:forKeys:count:].cold.5 + 0
	4   CoreFoundation                      0x00007fff204a51ff -[__NSPlaceholderDictionary initWithObjects:forKeys:count:] + 235
	5   CoreFoundation                      0x00007fff20438da3 +[NSDictionary dictionaryWithObjects:forKeys:count:] + 49
	6   Beacon                              0x0000000106e99a8e +[HSBeaconFontIcons attributedStringForIcon:color:size:] + 241
	7   Beacon                              0x0000000106e70904 -[HSBeaconAnswersSearchBar setCenteredPlaceholder:] + 243
	8   Beacon                              0x0000000106eba93d -[HSBeaconSuggestionsContainerController setupSearchBar] + 915
	9   Beacon                              0x0000000106eba123 -[HSBeaconSuggestionsContainerController viewDidLoad] + 229
	10  UIKitCore                           0x00007fff23f2c73b -[UIViewController _sendViewDidLoadWithAppearanceProxyObjectTaggingEnabled] + 88
	11  UIKitCore                           0x00007fff23f31022 -[UIViewController loadViewIfRequired] + 1084
	12  UIKitCore                           0x00007fff23e6800e -[UINavigationController _updateScrollViewFromViewController:toViewController:] + 162
	13  UIKitCore                           0x00007fff23e682f8 -[UINavigationController _startTransition:fromViewController:toViewController:] + 154
	14  UIKitCore                           0x00007fff23e69371 -[UINavigationController _startDeferredTransitionIfNeeded:] + 851
	15  UIKitCore                           0x00007fff23e6c131 -[UINavigationController pushViewController:transition:forceImmediate:] + 2046
	16  UIKitCore                           0x00007fff23f3f5c0 +[UIViewController _performWithoutDeferringTransitionsAllowingAnimation:actions:] + 153
	17  UIKitCore                           0x00007fff23e6b716 -[UINavigationController pushViewController:animated:] + 532
	18  Beacon                              0x0000000106ecb378 -[HSSuggestionsCoordinator start] + 813
	19  Beacon                              0x0000000106e67a33 __68-[HSBeacon loadNewBeacon:postLoadCallback:signature:viewController:]_block_invoke_2 + 344
	20  libdispatch.dylib                   0x000000010c55e8ac _dispatch_call_block_and_release + 12
	21  libdispatch.dylib                   0x000000010c55fa88 _dispatch_client_callout + 8
	22  libdispatch.dylib                   0x000000010c56df23 _dispatch_main_queue_callback_4CF + 1152
	23  CoreFoundation                      0x00007fff203a8276 __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 9
	24  CoreFoundation                      0x00007fff203a2b06 __CFRunLoopRun + 2685
	25  CoreFoundation                      0x00007fff203a1b9e CFRunLoopRunSpecific + 567
	26  GraphicsServices                    0x00007fff2b773db3 GSEventRunModal + 139
	27  UIKitCore                           0x00007fff24660af3 -[UIApplication _run] + 912
	28  UIKitCore                           0x00007fff24665a04 UIApplicationMain + 101
	29  Runner                              0x00000001054eb398 main + 72
	30  libdyld.dylib                       0x00007fff20257415 start + 1
)
libc++abi.dylib: terminating with uncaught exception of type NSException
*** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '*** -[__NSPlaceholderDictionary initWithObjects:forKeys:count:]: attempt to insert nil object from objects[0]'
terminating with uncaught exception of type NSException
```

**Comment by adrianne-helpscout** [comment]
@rafaelks I'm going to open a new issue for your new problem. Let's leave this issue for tracking assets not loading with CocoaPods. Thanks.

**Comment by ScottPetit** [comment]
Hey @AyushBherwani1998,

Yes it seems like your assumption is right.  This seems like a more extreme example of Beacon not being able to load assets that it expects to be there.  I wonder, were you ever able to manually integrate Beacon into your project?

**Comment by AyushBherwani1998** [comment]
Hey @ScottPetit,

I'm looking into ways to manually integrate Beacon in my project, I come Android background, and I'm not very familiar with Cocoapods. I did try to make a plugin for the same, but unfortunately same issue. I can upload the source code of plugin and instructions to run, if you want to dig into issue.

**Comment by ScottPetit** [comment]
Hey @AyushBherwani1998,

Yea sharing the source code of the plugin would be helpful.  Also just to understand your project a bit more, would your Flutter app being using both the iOS and Android SDKs or just the iOS?

**Comment by KelvinForteta** [comment]
Please this issue still exist, is there any solution to this?

**Comment by ScottPetit** [comment]
Hey @KelvinForteta,

Unfortunately we don't have any more info on why this issue happens in Flutter projects but not native integrations.  Are you using CocoaPods to integrate Beacon into your app?  If so, could you try integrating the [latest release](https://github.com/helpscout/beacon-ios-sdk/releases/tag/2.1.2) manually into your Xcode project?

Also another piece of helpful information would be if `Beacon.bundle` is anywhere in your Xcode project, either in the Beacon pod or in the `App.framework` generated by Flutter?

**Comment by KelvinForteta** [comment]
Yes I am using cocoaPods and I have the latest version 2.1.2.


![Screenshot 2021-01-09 at 09 04 40](https://user-images.githubusercontent.com/53996412/104086921-10f6a300-525c-11eb-8383-d517f7d5bc87.png)

**Comment by jotaemepereira** [comment]
@KelvinForteta Did you try what ScottPetit recommended on integrating the [latest release](https://github.com/helpscout/beacon-ios-sdk/releases/tag/2.1.2) manually into the project?

[Here](https://www.simpleswiftguide.com/how-to-add-xcframework-to-xcode-project/) is a quick guide on how to integrate a `xcframework` into a project. Let me know if that helps.

**Comment by ScottPetit** [comment]
Hey @KelvinForteta,

I just wanted to reach out because we made some updates to how resources are referenced via Cocoapods that could maybe resolve this issue.  You can access by updating your Podfile to point at

pod 'Beacon', :git => 'https://github.com/helpscout/beacon-ios-sdk.git', :commit => '1fd82f6'
or

pod 'Beacon', :git => 'https://github.com/helpscout/beacon-ios-sdk.git', :branch => 'master'
Once we release a new version you should be able to go back to just pod 'Beacon'.

Let us know if you're able to take a look at that and if it resolves your issue.

**Comment by KelvinForteta** [comment]
Ok thanks, I will try it out.

**Comment by shaunmcilroy** [comment]
Hey @KelvinForteta - Shaun from the mobile team here. I'm following up to see if Scott's suggestions above were of any help to you. Thanks!

**Comment by shaunmcilroy** [comment]
Hi @KelvinForteta ([link](https://github.com/helpscout/beacon-ios-sdk/issues/91#issuecomment-772595371)) & @AyushBherwani1998 ([link](https://github.com/helpscout/beacon-ios-sdk/issues/91#issuecomment-721231468))

Shaun here following up to our previous requests. Can you please confirm for us if this issue still affects your team, and if so did you follow our previous instructions to get Beacon for iOS loading assets correctly?

**Comment by ttyn618** [comment]
flutter project,  I am using cocoaPods and I have the latest version 2.1.2. I'm facing the similar issue.

**Comment by jotaemepereira** [comment]
Hey @ttyn618, are you using the latest Beacon version and tried what Scott suggested [here](https://github.com/helpscout/beacon-ios-sdk/issues/91#issuecomment-772590632)?

**Comment by ScottPetit** [comment]
Hey @ttyn618 just wanted to update you and other people that may be tracking this card.  After helping debug another issue we were seeing in a Flutter App (https://github.com/helpscout/beacon-ios-sdk/issues/161) we were able to reproduce this issue.

We're working on a fix and will update everyone here when it's available.

**Comment by ScottPetit** [comment]
👋  @ttyn618,

Just wanted to let you know that we just released version [2.2.2](https://github.com/helpscout/beacon-ios-sdk/blob/master/CHANGELOG.md) of the SDK that should resolve this issue.  Let us know if you run into any other issues.
